ntp-udp/src/lib.rs

```diff
@@ -1,219 +1,424 @@
-    }
+    mod set_timestamping_options {
+        use std::os::unix::prelude::AsRawFd;
 
-    if timestamping.tx_software {
-        // - we want send timestamps
-        // - return just the timestamp, don't send the full message along
-        // - tag the timestamp with an ID
-        options |= libc::SOF_TIMESTAMPING_TX_SOFTWARE
-            | libc::SOF_TIMESTAMPING_OPT_TSONLY
-            | libc::SOF_TIMESTAMPING_OPT_ID;
-    }
+        use super::{cerr, TimestampingConfig};
 
-    unsafe {
-        cerr(libc::setsockopt(
-            fd,
-            libc::SOL_SOCKET,
-            libc::SO_TIMESTAMPING,
-            &options as *const _ as *const libc::c_void,
-            std::mem::size_of::<libc::c_int>() as libc::socklen_t,
-        ))?
-    };
+        pub(crate) fn set_timestamping_options(
+            udp_socket: &std::net::UdpSocket,
+            timestamping: TimestampingConfig,
+        ) -> std::io::Result<()> {
+            let fd = udp_socket.as_raw_fd();
 
-    Ok(())
-}
+            let mut options = 0;
 
-/// Turn a C failure (-1 is returned) into a rust Result
-pub(crate) fn cerr(t: libc::c_int) -> std::io::Result<libc::c_int> {
-    match t {
-        -1 => Err(std::io::Error::last_os_error()),
-        _ => Ok(t),
-    }
-}
+            if timestamping.rx_software || timestamping.tx_software {
+                // enable software timestamping
+                options |= libc::SOF_TIMESTAMPING_SOFTWARE
+            }
+
+            if timestamping.rx_software {
+                // we want receive timestamps
+                options |= libc::SOF_TIMESTAMPING_RX_SOFTWARE
+            }
 
-/// Receive a message on a socket (retry if interrupted)
-fn receive_message(
-    socket: &std::net::UdpSocket,
-    message_header: &mut libc::msghdr,
-    flags: libc::c_int,
-) -> std::io::Result<libc::c_int> {
-    loop {
-        match cerr(unsafe { libc::recvmsg(socket.as_raw_fd(), message_header, flags) } as _) {
-            Err(e) if std::io::ErrorKind::Interrupted == e.kind() => {
-                // retry when the recv was interrupted
-                continue;
+            if timestamping.tx_software {
+                // - we want send timestamps
+                // - return just the timestamp, don't send the full message along
+                // - tag the timestamp with an ID
+                options |= libc::SOF_TIMESTAMPING_TX_SOFTWARE
+                    | libc::SOF_TIMESTAMPING_OPT_TSONLY
+                    | libc::SOF_TIMESTAMPING_OPT_ID;
             }
 
-            other => return other,
+            // for documentation on SO_TIMESTAMPING see
+            // https://www.kernel.org/doc/Documentation/networking/timestamping.txt
+            // Safety:
+            // we have a reference to the socket, so fd is a valid file descriptor for the duration of the call
+            // SOL_SOCKET + SO_TIMESTAMPING expect a *u32 as value, which &options is. Furthermore, we own options
```

Comment:
Added reference in the safety comment. Unfortunately this is not that well documented actually, so that section (and its code sample) really is all we base this on.